### PR TITLE
[flang] Lower MATMUL to type specific runtime calls.

### PR DESCRIPTION
**Pull request by vzakhari**
Lower MATMUL to the new runtime entries added in #97406.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
@vzakhari Did you see any benefits for 178.galgel/spec2000 ?

**Comment by vzakhari** [comment]
> @vzakhari Did you see any benefits for 178.galgel/spec2000 ?

Kiran, I did not measure it, because I do not expect much performance change.  I will keep an eye on our performance tracking, and will let you know if there is any noticeable change.

File: flang/include/flang/Optimizer/Support/Utils.h [hunk 1]
```diff
@@ -133,10 +142,70 @@ inline void intrinsicTypeTODO(fir::FirOpBuilder &builder, mlir::Type type,
...
+  else if (type == fir::ComplexType::get(loc.getContext(), 16))
+    return {Fortran::common::TypeCategory::Complex, 16};
+  else if (type == fir::LogicalType::get(loc.getContext(), 1))
+    return {Fortran::common::TypeCategory::Logical, 1};
+  else if (type == fir::LogicalType::get(loc.getContext(), 2))
+    return {Fortran::common::TypeCategory::Logical, 2};
+  else if (type == fir::LogicalType::get(loc.getContext(), 4))
+    return {Fortran::common::TypeCategory::Logical, 4};
+  else if (type == fir::LogicalType::get(loc.getContext(), 8))
+    return {Fortran::common::TypeCategory::Logical, 8};
```
Comment (jeanPerier) [inline flang/include/flang/Optimizer/Support/Utils.h:203]: nit:
```
if(complexType = mlir::dyn_cast<fir::ComplexType>(type))
  return {Fortran::common::TypeCategory::Complex, complexType.getFKind()};
else if (logicalType = mlir::dyn_cast<fir::LogicalType>(type))
  return {Fortran::common::TypeCategory::Logical, logicalType.getFKind()};
```

**Review by jeanPerier — APPROVED** [review]
Thanks Slava!

**Pull request merged**